*** /usr/local/java/jdk1.4/src/java/io/ObjectInputStream.java	Sat Jan  5 03:13:50 2002
--- ObjectInputStream.java	Tue Feb 19 10:58:20 2002
***************
*** 1,3 ****
--- 1,8 ----
+ /** Modified by cougaar
+  * <copyright>
+  * </copyright>
+  **/
+ 
  /*
   * @(#)ObjectInputStream.java	1.136 01/12/03
   *
***************
*** 1618,1624 ****
  
  	Object obj;
  	try {
! 	    obj = desc.isInstantiable() ? desc.newInstance() : null;
  	} catch (Exception ex) {
  	    throw new InvalidClassException(
  		desc.forClass().getName(), "unable to create instance");
--- 1623,1631 ----
  
  	Object obj;
  	try {
!           // obj = desc.isInstantiable() ? desc.newInstance() : null;
!           // COUGAAR mod
!           obj = desc.isInstantiable() ? newInstanceFromDesc(desc) : null;
  	} catch (Exception ex) {
  	    throw new InvalidClassException(
  		desc.forClass().getName(), "unable to create instance");
***************
*** 3325,3328 ****
--- 3332,3343 ----
  	    }
  	}
      }
+ 
+   // COUGAAR
+   protected Object newInstanceFromDesc(ObjectStreamClass desc) 
+     throws InstantiationException, java.lang.reflect.InvocationTargetException
+   {
+     return desc.newInstance();
+   }
  }
+ 
